File: backend/src/main/java/com/berksprojects/expensetracker/service/expense/ExpenseService.java
package com.berksprojects.expensetracker.service.expense;

import com.berksprojects.expensetracker.dto.ExpenseDto;
import com.berksprojects.expensetracker.entity.Expense;

import java.util.List;

public interface ExpenseService {
    Expense postExpense(ExpenseDto expenseDto);

    List<Expense> getAllExpenses();

    Expense getExpenseById(Long id);

    Expense updateExpense(Long id, ExpenseDto expenseDto);

    void deleteExpense(Long id);
}

File: frontend/src/app/services/expense/expense.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

const BASIC_URL = "http://localhost:8080/"

@Injectable({
  providedIn: 'root'
})
export class ExpenseService {

  constructor(private http: HttpClient) { }

  getExpenseById(id:number): Observable<any> {
    return this.http.get(BASIC_URL + "api/expense/" + id);
  }

  postExpense(expenseDto:any): Observable<any> {
    return this.http.post(BASIC_URL + "api/expense", expenseDto);
  }

  updateExpense(id:number, expenseDto:any): Observable<any> {
    return this.http.put(BASIC_URL + "api/expense/" + id, expenseDto);
  }

  getAllExpenses(): Observable<any> {
    return this.http.get(BASIC_URL + "api/expense/all");
  }

  deleteExpense(expenseId:number): Observable<any> {
    return this.http.delete(BASIC_URL + "api/expense/" + expenseId);
  }
}
